system/lib/wasmfs/backends/opfs_backend.cpp

```diff
@@ -0,0 +1,279 @@
+    : Directory(mode, backend), dirID(dirID), proxy(proxy) {}
+
+  ~OPFSDirectory() override {
+    // Never free the root directory ID.
+    if (dirID != 0) {
+      proxy([&]() { _wasmfs_opfs_free_directory(dirID); });
+    }
+  }
+
+private:
+  std::shared_ptr<File> getChild(const std::string& name) override {
+    int childType = 0, childID = 0;
+    proxy([&](auto ctx) {
+      _wasmfs_opfs_get_child(
+        ctx.ctx, dirID, name.c_str(), &childType, &childID);
+    });
+    if (childID < 0) {
+      // TODO: More fine-grained error reporting.
+      return NULL;
+    }
+    if (childType == 1) {
+      return std::make_shared<OPFSFile>(0777, getBackend(), childID, proxy);
+    } else if (childType == 2) {
+      return std::make_shared<OPFSDirectory>(
+        0777, getBackend(), childID, proxy);
+    } else {
+      WASMFS_UNREACHABLE("Unexpected child type");
+    }
+  }
+
+  std::shared_ptr<DataFile> insertDataFile(const std::string& name,
+                                           mode_t mode) override {
+    int childID = 0;
+    proxy([&](auto ctx) {
+      _wasmfs_opfs_insert_file(ctx.ctx, dirID, name.c_str(), &childID);
+    });
+    // TODO: Handle errors gracefully.
+    assert(childID >= 0);
+    return std::make_shared<OPFSFile>(mode, getBackend(), childID, proxy);
+  }
+
+  std::shared_ptr<Directory> insertDirectory(const std::string& name,
+                                             mode_t mode) override {
+    int childID = 0;
+    proxy([&](auto ctx) {
+      _wasmfs_opfs_insert_directory(ctx.ctx, dirID, name.c_str(), &childID);
+    });
+    // TODO: Handle errors gracefully.
+    assert(childID >= 0);
+    return std::make_shared<OPFSDirectory>(mode, getBackend(), childID, proxy);
+  }
+
+  std::shared_ptr<Symlink> insertSymlink(const std::string& name,
+                                         const std::string& target) override {
+    // Symlinks not supported.
+    return nullptr;
+  }
+
+  bool insertMove(const std::string& name,
+                  std::shared_ptr<File> file) override {
+    auto old_file = std::static_pointer_cast<OPFSFile>(file);
+    proxy([&](auto ctx) {
+      _wasmfs_opfs_move(ctx.ctx, old_file->fileID, dirID, name.c_str());
+    });
+    // TODO: Handle errors.
+    return true;
+  }
+
+  bool removeChild(const std::string& name) override {
+    proxy([&](auto ctx) {
+      _wasmfs_opfs_remove_child(ctx.ctx, dirID, name.c_str());
+    });
+    return true;
+  }
+
+  size_t getNumEntries() override { return getEntries().size(); }
+
+  std::vector<Directory::Entry> getEntries() override {
+    std::vector<Directory::Entry> entries;
+    proxy(
+      [&](auto ctx) { _wasmfs_opfs_get_entries(ctx.ctx, dirID, &entries); });
+    return entries;
+  }
+};
+
+class OPFSBackend : public Backend {
+public:
+  ProxyWorker proxy;
+
+  std::shared_ptr<DataFile> createFile(mode_t mode) override {
+    // No way to support a raw file without a parent directory.
+    return nullptr;
```

Comment:
Should this abort? I'm unsure what it means to return nullptr here - is that meant to signal an error? I don't see an explanation of the API in `system/lib/wasmfs/backend.h` so maybe we should describe it there.

Reply:
Ideally this would signal an error, but yeah, that's not documented and probably not handled correctly either. I'll add a TODO since I would like to avoid changes to the core system in this PR.